In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = r"C:\Users\fardi\OneDrive\Desktop\CF\Data Immersion\Python Fundamentals for Data Analysts\06-09-2021 Instacart Basket Analysis"

In [3]:
# Loading customer dataset
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
df_cust.shape

(206209, 10)

In [6]:
df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [12]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Data Wrangling

#### rename columns names for consistency

In [13]:
df_cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [14]:
df_cust.rename(columns = {'Surnam' : 'family_name'}, inplace = True)

In [15]:
df_cust.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [16]:
df_cust.rename(columns = {'STATE' : 'state'}, inplace = True)

In [17]:
df_cust.rename(columns = {'Age' : 'age'}, inplace = True)

In [18]:
df_cust.rename(columns = {'n_dependants' : 'number_of_dependents'}, inplace = True)

In [19]:
df_cust.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [20]:
df_cust.head()

user_id first_name family_name  gender       state  age date_joined  \
0   26711    Deborah    Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia        Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth      Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle       Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann     Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

###  Data quality and Consistency checks

In [21]:
# finding missing values
df_cust.isnull().sum()

user_id                     0
first_name              11259
family_name                 0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
family_status               0
income                      0
dtype: int64

we have 11259 missing first_names, but before dealing with the missing data, I wanted to make sure if I need the first_name column for my analysis or I can drop the column completely

In [23]:
# create a subset of the dataframe containing only missing values
df_nan = df_cust[df_cust['first_name'].isnull()==True]

In [24]:
df_nan

user_id first_name family_name  gender         state  age date_joined  \
53       76659        NaN     Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN       Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN      Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN     Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN      Dawson  Female          Utah   63    1/1/2017   
...        ...        ...         ...     ...           ...  ...         ...   
206038  121317        NaN      Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN    Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN       Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN       Floyd  Female    California   39    4/1/2020   
206171  116898        NaN     Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependents family_status  income  
53                         2       married   41709  
73                         0        single   82518  
82                         3       married  117099  
99                         1       married  155673  
105                        3       married  151819  
...                      ...           ...     ...  
206038                     3       married   87783  
206044                     2       married  108488  
206090                     1       married   45275  
206162                     0        single   56325  
206171                     2       married   59222  

[11259 rows x 10 columns]

In [25]:
# finding duplicates
df_dups = df_cust[df_cust.duplicated()]

In [26]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, family_name, gender, state, age, date_joined, number_of_dependents, family_status, income]
Index: []

There are no duplicates in the dataset

In [27]:
# Drop column not needed
df_cust = df_cust.drop('first_name',1)

In [28]:
df_cust = df_cust.drop('date_joined',1)

In [29]:
df_cust.shape

(206209, 8)

In [30]:
df_cust.head()

user_id family_name  gender       state  age  number_of_dependents  \
0   26711    Esquivel  Female    Missouri   48                     3   
1   33890        Hart  Female  New Mexico   36                     0   
2   65803      Farley    Male       Idaho   35                     2   
3  125935       Hicks  Female        Iowa   40                     0   
4  130797     Gilmore  Female    Maryland   26                     1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

### Merging datatset

In [36]:
# Loading main dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_complete.pkl'))

In [38]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy      Least busy   
2         Soda  ...  Mid-range product  Regularly busy      Least busy   
3         Soda  ...  Mid-range product      Least busy      Least busy   
4         Soda  ...  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  loyalty_flag mean_price spender_flag  \
0        Average orders        10  New customer   6.367797  Low spender   
1        Average orders        10  New customer   6.367797  Low spender   
2           Most orders        10  New customer   6.367797  Low spender   
3        Average orders        10  New customer   6.367797  Low spender   
4           Most orders        10  New customer   6.367797  Low spender   

   median_order    frequency_flag  
0          20.0  Regular customer  
1          20.0  Regular customer  
2          20.0  Regular customer  
3          20.0  Regular customer  
4          20.0  Regular customer  

[5 rows x 24 columns]

In [39]:
ords_prods_merge.shape

(32404859, 24)

In [40]:
df_cust.shape

(206209, 8)

In [42]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  mean_price              float64 
 21  spende

In [46]:
ords_prods_merge = ords_prods_merge.drop('_merge',1)

In [47]:
orders_products_all = ords_prods_merge.merge(df_cust, on = ['user_id'], how = 'inner',indicator = True)

In [49]:
# checking if full merge
orders_products_all['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [50]:
orders_products_all.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  median_order    frequency_flag  family_name  gender  \
0         Soda  ...          20.0  Regular customer       Nguyen  Female   
1         Soda  ...          20.0  Regular customer       Nguyen  Female   
2         Soda  ...          20.0  Regular customer       Nguyen  Female   
3         Soda  ...          20.0  Regular customer       Nguyen  Female   
4         Soda  ...          20.0  Regular customer       Nguyen  Female   

     state age number_of_dependents  family_status income  _merge  
0  Alabama  31                    3        married  40423    both  
1  Alabama  31                    3        married  40423    both  
2  Alabama  31                    3        married  40423    both  
3  Alabama  31                    3        married  40423    both  
4  Alabama  31                    3        married  40423    both  

[5 rows x 31 columns]

In [53]:
orders_products_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 30 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  price_range_loc         object 
 14  busiest_day             object 
 15  busiest_days            object 
 16  busiest_period_of_day   object 
 17  max_order               int64  
 18  loyalty_flag            object 
 19  mean_price              float64
 20  spender_flag            object 
 21  median_order            float

In [51]:
orders_products_all = orders_products_all.drop('_merge',1)

### Export dataset

In [52]:
orders_products_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))